File: ios/Reb/Models/Protocol.swift
import Foundation

// MARK: - Protocol Message Types (mirrors server/src/protocol.ts)

enum MessageType: String, Codable {
    case auth
    case authResult = "auth_result"
    case createSession = "create_session"
    case sessionCreated = "session_created"
    case resize
    case listSessions = "list_sessions"
    case sessionList = "session_list"
    case killSession = "kill_session"
    case sessionEnded = "session_ended"
    case error
    case output
    case input
}

// MARK: - Client → Server

struct AuthMessage: Codable {
    let type = "auth"
    let token: String
}

struct CreateSessionMessage: Codable {
    let type = "create_session"
    let cols: Int
    let rows: Int
    var command: String?
}

struct ResizeMessage: Codable {
    let type = "resize"
    let sessionId: String
    let cols: Int
    let rows: Int
}

struct ListSessionsMessage: Codable {
    let type = "list_sessions"
}

struct KillSessionMessage: Codable {
    let type = "kill_session"
    let sessionId: String
}

// MARK: - Server → Client

struct AuthResultPayload: Codable {
    let type: String
    let success: Bool
    let error: String?
}

struct SessionCreatedPayload: Codable {
    let type: String
    let sessionId: String
}

struct SessionInfo: Codable, Identifiable {
    let id: String
    let pid: Int
    let createdAt: String
    let command: String
}

struct SessionListPayload: Codable {
    let type: String
    let sessions: [SessionInfo]
}

struct SessionEndedPayload: Codable {
    let type: String
    let sessionId: String
    let exitCode: Int
}

struct ErrorPayload: Codable {
    let type: String
    let message: String
    let context: String?
}

// MARK: - Generic envelope for routing

struct MessageEnvelope: Codable {
    let type: String
}

// MARK: - Binary Frame Constants

let binaryInputTag: UInt8 = 0x01
let binaryOutputTag: UInt8 = 0x02

func uuidToBytes(_ uuid: String) -> Data {
    let hex = uuid.replacingOccurrences(of: "-", with: "")
    var data = Data()
    var index = hex.startIndex
    for _ in 0..<16 {
        let nextIndex = hex.index(index, offsetBy: 2)
        if let byte = UInt8(hex[index..<nextIndex], radix: 16) {
            data.append(byte)
        }
        index = nextIndex
    }
    return data
}

func bytesToUuid(_ data: Data, offset: Int = 0) -> String {
    let hex = data[offset..<(offset + 16)].map { String(format: "%02x", $0) }.joined()
    let parts = [
        String(hex.prefix(8)),
        String(hex.dropFirst(8).prefix(4)),
        String(hex.dropFirst(12).prefix(4)),
        String(hex.dropFirst(16).prefix(4)),
        String(hex.dropFirst(20)),
    ]
    return parts.joined(separator: "-")
}

func buildInputFrame(sessionId: String, data: Data) -> Data {
    var frame = Data([binaryInputTag])
    frame.append(uuidToBytes(sessionId))
    frame.append(data)
    return frame
}

func parseOutputFrame(_ frame: Data) -> (sessionId: String, data: Data)? {
    guard frame.count >= 17, frame[0] == binaryOutputTag else { return nil }
    let sessionId = bytesToUuid(frame, offset: 1)
    let data = frame.subdata(in: 17..<frame.count)
    return (sessionId, data)
}

File: server/src/protocol.ts
// ============================================================
// Reb Protocol — WebSocket message types between iOS ↔ Server
// ============================================================

// --- Client → Server ---

export interface AuthMessage {
  type: 'auth';
  token: string;
}

export interface CreateSessionMessage {
  type: 'create_session';
  cols: number;
  rows: number;
  cli?: string;     // 'claude' | 'copilot' — resolved server-side
  command?: string;  // custom command override
}

export interface InputMessage {
  type: 'input';
  sessionId: string;
  // data is sent as a separate binary frame immediately after this JSON frame
}

export interface ResizeMessage {
  type: 'resize';
  sessionId: string;
  cols: number;
  rows: number;
}

export interface ListSessionsMessage {
  type: 'list_sessions';
}

export interface KillSessionMessage {
  type: 'kill_session';
  sessionId: string;
}

export type ClientMessage =
  | AuthMessage
  | CreateSessionMessage
  | InputMessage
  | ResizeMessage
  | ListSessionsMessage
  | KillSessionMessage;

// --- Server → Client ---

export interface AuthResultMessage {
  type: 'auth_result';
  success: boolean;
  error?: string;
}

export interface SessionCreatedMessage {
  type: 'session_created';
  sessionId: string;
  command: string;
}

export interface OutputMessage {
  type: 'output';
  sessionId: string;
  // data is sent as a separate binary frame immediately after this JSON frame
}

export interface SessionListMessage {
  type: 'session_list';
  sessions: SessionInfo[];
}

export interface SessionInfo {
  id: string;
  pid: number;
  createdAt: string;
  command: string;
}

export interface SessionEndedMessage {
  type: 'session_ended';
  sessionId: string;
  exitCode: number;
}

export interface ErrorMessage {
  type: 'error';
  message: string;
  context?: string;
}

export type ServerMessage =
  | AuthResultMessage
  | SessionCreatedMessage
  | OutputMessage
  | SessionListMessage
  | SessionEndedMessage
  | ErrorMessage;

// --- Binary Frame Protocol ---
// For efficiency, terminal I/O uses a binary frame format:
//
// Client → Server (input):
//   [1 byte: 0x01] [16 bytes: sessionId as UUID] [rest: terminal data]
//
// Server → Client (output):
//   [1 byte: 0x02] [16 bytes: sessionId as UUID] [rest: terminal data]

export const BINARY_INPUT_TAG = 0x01;
export const BINARY_OUTPUT_TAG = 0x02;

// Helper: encode UUID string to 16-byte buffer
export function uuidToBytes(uuid: string): Buffer {
  return Buffer.from(uuid.replace(/-/g, ''), 'hex');
}

// Helper: decode 16-byte buffer to UUID string
export function bytesToUuid(buf: Buffer, offset: number = 0): string {
  const hex = buf.subarray(offset, offset + 16).toString('hex');
  return [
    hex.slice(0, 8),
    hex.slice(8, 12),
    hex.slice(12, 16),
    hex.slice(16, 20),
    hex.slice(20, 32),
  ].join('-');
}

// Build a binary output frame
export function buildOutputFrame(sessionId: string, data: Buffer): Buffer {
  const header = Buffer.alloc(17);
  header[0] = BINARY_OUTPUT_TAG;
  uuidToBytes(sessionId).copy(header, 1);
  return Buffer.concat([header, data]);
}

// Parse a binary input frame
export function parseInputFrame(frame: Buffer): { sessionId: string; data: Buffer } | null {
  if (frame.length < 17 || frame[0] !== BINARY_INPUT_TAG) return null;
  return {
    sessionId: bytesToUuid(frame, 1),
    data: frame.subarray(17),
  };
}
